core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -643,6 +643,275 @@ class WorkspaceService(
+        GoogleProject(workspace.googleProjectId.value),
+        oldBillingProjectOwnerPolicyEmail,
+        IamMemberTypes.Group,
+        Set(terraBillingProjectOwnerRole, terraWorkspaceCanComputeRole, terraWorkspaceNextflowRole),
+        retryIfGroupDoesNotExist = true
+      )
+
+      // Change billing account
+      _ <- gcsDAO.setBillingAccount(workspace.googleProjectId, destBillingAccountName, ctx.toTracingContext)
+    } yield ()
+
+  private def rollbackUpdateWorkspaceBillingInGCP(workspace: Workspace,
+                                                  oldBillingProjectOwnerPolicyEmail: WorkbenchEmail,
+                                                  newBillingProjectOwnerPolicyEmail: WorkbenchEmail
+  ): Future[Unit] =
+    updateWorkspaceBillingInGCP(workspace,
+                                workspace.currentBillingAccountOnGoogleProject,
+                                newBillingProjectOwnerPolicyEmail,
+                                oldBillingProjectOwnerPolicyEmail
+    ).recover { case rollbackEx =>
+      logger.warn(s"Failed to roll back GCP billing update for workspace ${workspace.briefName}", rollbackEx)
+    }
+
+  private def updateWorkspaceBillingInSam(workspace: Workspace,
+                                          oldBillingProjectOwnerPolicyEmail: WorkbenchEmail,
+                                          newBillingProjectOwnerPolicyEmail: WorkbenchEmail
+  ): Future[Unit] =
+    for {
+      // Add new billing project owner email to workspace owner policy and remove old owner
+      _ <- samDAO.addUserToPolicy(
+        SamResourceTypeNames.workspace,
+        workspace.workspaceId,
+        SamWorkspacePolicyNames.projectOwner,
+        newBillingProjectOwnerPolicyEmail.value,
+        samDAO.rawlsSAContext
+      )
+      _ <- samDAO.removeUserFromPolicy(
+        SamResourceTypeNames.workspace,
+        workspace.workspaceId,
+        SamWorkspacePolicyNames.projectOwner,
+        oldBillingProjectOwnerPolicyEmail.value,
+        samDAO.rawlsSAContext
+      )
+
+      // Remove and re-add fastpass grants
+      _ <- fastPassServiceConstructor(ctx).removeFastPassGrantsForWorkspace(workspace)
+      _ <- fastPassServiceConstructor(ctx).syncFastPassesForUserInWorkspace(workspace)
+    } yield ()
+
+  private def rollbackUpdateWorkspaceBillingInSam(workspace: Workspace,
+                                                  oldBillingProjectOwnerPolicyEmail: WorkbenchEmail,
+                                                  newBillingProjectOwnerPolicyEmail: WorkbenchEmail
+  ): Future[Unit] =
+    updateWorkspaceBillingInSam(workspace, newBillingProjectOwnerPolicyEmail, oldBillingProjectOwnerPolicyEmail)
+      .recover { case rollbackEx =>
+        logger.warn(s"Failed to roll back SAM billing update for workspace ${workspace.briefName}", rollbackEx)
+      }
+
+  def updateWorkspaceBilling(workspace: Workspace, destBillingProject: RawlsBillingProject): Future[Option[Workspace]] =
```

Comment:
optional, up to you: for readability and organization, consider making a new class to hold the update and rollback methods for GCP and Sam. If that new class accepted the workspace, new billing account name, old policy email, and new policy email as parameters to its constructor, then the update/rollback methods wouldn't need to accept any parameters and this `updateWorkspaceBilling` would probably read a lot easier.

Reply:
I like the suggestion! I am going to skip it though in the interest of time

Reply:
I like the suggestion! I am going to skip it though in the interest of time